From d1eba132bb5823eacc0f01acdb7467753a386129 Mon Sep 17 00:00:00 2001
From: Pinyao Ting <pinyaoting@google.com>
Date: Thu, 8 Jun 2023 14:38:19 -0700
Subject: [PATCH] Restrict number of shortcuts can be added through
 addDynamicShortcuts

This CL fixes the issue where, when an app have multiple main
activities, the total number of shortcuts can grow indefinitely if they
were published through addDynamicShortcuts.

Bug: 281061287
Test: manual
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:2d93aabdc4905b36ee684533904029cfc61533b7)
Merged-In: Ib3eecefee34517b670c59dd5b8526fe9eb24f463
Change-Id: Ib3eecefee34517b670c59dd5b8526fe9eb24f463
---
 services/core/java/com/android/server/pm/ShortcutPackage.java | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/pm/ShortcutPackage.java b/services/core/java/com/android/server/pm/ShortcutPackage.java
index b2d78ec7d95d..4bc3cdb730a3 100644
--- a/services/core/java/com/android/server/pm/ShortcutPackage.java
+++ b/services/core/java/com/android/server/pm/ShortcutPackage.java
@@ -305,6 +305,7 @@ class ShortcutPackage extends ShortcutPackageItem {
         // Extract Icon and update the icon res ID and the bitmap path.
         s.saveIconAndFixUpShortcutLocked(newShortcut);
         s.fixUpShortcutResourceNamesAndValues(newShortcut);
+        ensureShortcutCountBeforePush();
         mShortcuts.put(newShortcut.getId(), newShortcut);
     }
 
@@ -351,7 +352,6 @@ class ShortcutPackage extends ShortcutPackageItem {
             @NonNull List<ShortcutInfo> changedShortcuts) {
         Preconditions.checkArgument(newShortcut.isEnabled(),
                 "pushDynamicShortcuts() cannot publish disabled shortcuts");
-        ensureShortcutCountBeforePush();
 
         newShortcut.addFlags(ShortcutInfo.FLAG_DYNAMIC);
 
-- 
2.42.0.820.g83a721a137-goog

